packages/regl-worldview/src/WorldviewContext.js

```diff
@@ -240,63 +244,104 @@ export class WorldviewContext {
-          // if the mouse quickly leaves the draw area during a read operation
-          // reading outside the bounds of the framebuffer causes errors
-          // and puts regl into a bad internal state.
-          // https://github.com/regl-project/regl/blob/28fbf71c871498c608d9ec741d47e34d44af0eb5/lib/read.js#L57
-          if (x < Math.floor(width) && y < Math.floor(height) && x >= 0 && y >= 0) {
-            const pixel = new Uint8Array(4);
-
-            // read pixel value from the frame buffer
-            regl.read({
-              x,
-              y,
-              width: 1,
-              height: 1,
-              data: pixel,
-            });
-
-            objectId = getIdFromColor(pixel);
-          }
-          resolve(objectId);
+          do {
+            if (counter === MAX_NUMBER_OF_LAYERS) {
+              // Provide a max number of layers so this while loop doesn't crash the page.
+              console.error(
+                `Hit ${MAX_NUMBER_OF_LAYERS} iterations. There is either that number of rendered layers or a bug with hitmap events.`
+              );
+              break;
+            } else if (counter > 0 && !enableStackedObjectEvents) {
+              // Only allow multiple object selection if we've explicitly enabled it.
+              break;
+            }
+            counter++;
+            regl.clear({ color: [0, 0, 0, 1], depth: 1 });
+            this._drawInput(true, seenObjects);
+
+            // it's possible to get x/y values outside the framebuffer size
+            // if the mouse quickly leaves the draw area during a read operation
+            // reading outside the bounds of the framebuffer causes errors
+            // and puts regl into a bad internal state.
+            // https://github.com/regl-project/regl/blob/28fbf71c871498c608d9ec741d47e34d44af0eb5/lib/read.js#L57
+            if (x < Math.floor(width) && y < Math.floor(height) && x >= 0 && y >= 0) {
+              const pixel = new Uint8Array(4);
+
+              // read pixel value from the frame buffer
+              regl.read({
+                x,
+                y,
+                width: 1,
+                height: 1,
+                data: pixel,
+              });
+
+              currentObjectId = getIdFromColor(pixel);
+              if (currentObjectId && this.getDrawPropByHitmapId(currentObjectId).object) {
+                seenIds.push(currentObjectId);
+                seenObjects.push(this.getDrawPropByHitmapId(currentObjectId));
+              }
+            }
+          } while (currentObjectId !== 0);
+          resolve(seenIds);
         });
       });
     });
   }
 
-  callComponentHandlers = (objectId: number, ray: Ray, e: MouseEvent, mouseEventName: MouseEventEnum) => {
-    this._hitmapCalls.forEach((_, component) => {
+  callComponentHandlers = (hitmapIds: Array<HitmapId>, ray: Ray, e: MouseEvent, mouseEventName: MouseEventEnum) => {
```

Comment:
It feels a bit roundabout for `<Worldview>` to call this. Why not call this internally from `readHitmap` (need to rename probably though)? Then we can probably find the appropriate `component` more easily anyway.

Reply:
Not super big deal though, maybe the concerns are split kinda nicely now? 🤷‍♂

Reply:
Did this in a different way